src/activation.rs

```diff
@@ -182,13 +183,18 @@ pub fn receive_descriptors_with_names(
     Ok(out)
 }
 
-fn socks_from_fds(num_fds: usize) -> Result<Vec<FileDescriptor>, SdError> {
-    let mut descriptors = Vec::with_capacity(num_fds);
-    for fd_offset in 0..num_fds {
-        let index = SD_LISTEN_FDS_START
+fn socks_from_fds(fd_count: usize) -> Result<Vec<FileDescriptor>, SdError> {
+    let mut descriptors = Vec::with_capacity(fd_count);
+    for fd_offset in 0..fd_count {
+        let fd_num = SD_LISTEN_FDS_START
             .checked_add(fd_offset as i32)
-            .with_context(|| format!("overlarge file descriptor index: {}", num_fds))?;
-        let fd = FileDescriptor::try_from(index).unwrap_or_else(|(msg, val)| {
+            .with_context(|| format!("overlarge file descriptor index: {}", fd_count))?;
+        // Set CLOEXEC on the file descriptors we receive so that they aren't
+        // passed to programs exec'd from here, just like sd_listen_fds does.
+        if let Err(errno) = fcntl(fd_num, F_SETFD(FdFlag::FD_CLOEXEC)) {
+            log::warn!("couldn't set FD_CLOEXEC on {fd_num}: {errno}");
```

Comment:
I think we should stop and bubble-up the error here. That's how the upstream C implementation behaves (currently, as of v257).

Reply:
Done. I had been looking at the [implementation from the `systemd_socket` crate](https://github.com/Kixunil/systemd_socket/blob/7f6d85c459598f3f3077dc0956065def87d7e3e7/src/lib.rs#L188C1-L199C14), which treats it as a recoverable error. I think there’s merits to either approach but following systemd’s own example is probably better for this library.